kyaml/kio/byteio_reader.go

```diff
@@ -282,3 +299,46 @@ func (r *ByteReader) decode(index int, decoder *yaml.Decoder) (*yaml.RNode, erro
 	}
 	return yaml.NewRNode(node), nil
 }
+
+// seqIndentAnno derives the sequence indentation annotation value for the resource,
+// originalYAML is the input yaml string and node is the decoded equivalent of originalYAML,
+// the annotation value is decided by deriving the existing sequence indentation of resource
+func seqIndentAnno(node *yaml.Node, originalYAML string) string {
+	rNode := &yaml.RNode{}
+	rNode.SetYNode(node)
+	// step 1: derive the sequence indentation of the node
+	anno := rNode.GetAnnotations()
+	if anno[kioutil.SeqIndentAnnotation] != "" {
+		// the annotation already exists, so don't change it
+		return ""
+	}
+
+	// marshal the node with 2 space sequence indentation and calculate the diff
+	out, err := yaml.MarshalWithOptions(rNode.Document(), &yaml.EncoderOptions{
+		MapIndent: yaml.DefaultMapIndent,
+		SeqIndent: yaml.WideSeqIndent,
+	})
+	if err != nil {
+		return ""
+	}
+	twoSpaceIndentDiff := copyutil.PrettyFileDiff(string(out), originalYAML)
+
+	// marshal the node with 0 space sequence indentation and calculate the diff
+	out, err = yaml.MarshalWithOptions(rNode.Document(), &yaml.EncoderOptions{
+		MapIndent: yaml.DefaultMapIndent,
+		SeqIndent: yaml.CompactSeqIndent,
+	})
+	if err != nil {
+		return ""
+	}
+	noIndentDiff := copyutil.PrettyFileDiff(string(out), originalYAML)
+
+	var style string
+	if len(noIndentDiff) <= len(twoSpaceIndentDiff) {
```

Comment:
Will this heuristic be confounded by a non-2 general indent level with compact style? Rather than attempting to determine the average indent level, could we sample the first sequence somehow (scanning? column data on YNodes?) and use that? I imagine most documents would be consistent, and when they're not, "first" behaviour might be easier for the end user to understand than "average".

Reply:
From the feedback we received, users want to see less diffs. The goal here is to minimize the diff. So the "first" behavior doesn't achieve that goal.

And identifying first sequence is difficult problem. There would be many edge cases.

Reply:
Not sure on the complexity of identify the "first" behavior, but picking the "first" behavior has more benefits over the "average" when indentation is not consistent within a resource. It makes it more deterministic and can *potentially* make this function cheaper as well. We should definitely explore if it is a possibility.

Reply:
I did a benchmark analysis and found that the latency is 2.5 times more. Implemented "first" approach. There is no significant performance degradation. Since the new approach involves raw YAML string parsing, I might be missing few corner cases. We can improve it as we go.